logstash-core/src/main/java/org/logstash/instrument/witness/schedule/WitnessScheduler.java

```diff
@@ -0,0 +1,82 @@
+package org.logstash.instrument.witness.schedule;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Schedules {@link ScheduledWitness} to refresh themselves on an interval.
+ */
+public class WitnessScheduler {
+
+    private final ScheduledWitness witness;
+    private final ScheduledExecutorService executorService;
+    private static final Logger LOGGER = LogManager.getLogger(WitnessScheduler.class);
+
+    /**
+     * Constructor
+     *
+     * @param witness the {@link ScheduledWitness} to schedule
+     */
+    public WitnessScheduler(ScheduledWitness witness) {
+        this.witness = witness;
+
+        this.executorService = Executors.newScheduledThreadPool(1, ((Runnable r) -> {
+            Thread t = new Thread(r);
+            //Allow this thread to simply die when the JVM dies
+            t.setDaemon(true);
+            //Set the name
+            t.setName(witness.threadName());
+            //Set the priority
+            t.setPriority(witness.priority());
+            return t;
+        }));
+    }
+
+    /**
+     * Schedules the witness to refresh on provided schedule. Note - this implementation does not allow refreshes faster then every 1 second.
+     */
+    public void schedule() {
+        executorService.scheduleAtFixedRate(new RefreshRunnable(), 0, witness.every().getSeconds(), TimeUnit.SECONDS);
+    }
+
+    /**
+     * Shuts down the underlying executor service. Since these are daemon threads, this is not absolutely necessary.
+     */
+    public void shutdown(){
+        executorService.shutdown();
+        try {
+            if(!executorService.awaitTermination(5, TimeUnit.SECONDS)){
+                executorService.shutdownNow();
+            }
+        } catch (InterruptedException e) {
+            //do nothing
+        }
+    }
+
+    /**
+     * Runnable that will won't cancel the scheduled tasks on refresh if an exception is thrown, and throttles the log message.
+     */
+    class RefreshRunnable implements Runnable {
+
+       long lastLogged = 0;
+
+        @Override
+        public void run() {
+            try {
+                witness.refresh();
+            } catch (Exception e) {
```

Comment:
@jakelandis what exceptions do we assume to get here? Shouldn't we simply crash LS here on unexpected bugs instead of simply continuing to run with frozen metrics?

Reply:
The real world example I was thinking of was https://github.com/elastic/logstash/issues/6824 where CPU metrics could not be gathered on specific OS's.  I don't think the metrics should take down the primary function of Logstash if they don't work or there is a transient error.

Reply:
@jakelandis from practical experience I'd strongly disagree here. I think one of if not the main use case for the metrics is monitoring. If a malfunctioning LS may simply cause the metrics to freeze, this is a huge issue for users.
Errors like the one you linked we should handle on a case by case basis if we know that ignoring them is not affecting functionality overall, but suppressing `100%` of errors and freezing metrics in time makes them invalid for monitoring purposes imo.

Reply:
@original-brownbear - we may need a 3rd party to resolve this one, as I strongly agree with the approach here.  

* Not everyone using Logstash cares about metrics
* Metrics are a side effect of the main purpose, not an integral part of it's function
* There are known issues on some platforms, and crashing Logstash would dis-allow those platforms/configurations to update.
* This change as-is is passive to the prior version.

Reply:
@jakelandis 

> This change as-is is passive to the prior version.

I still feel like I'm right but ... you win with this one. I should've known ... even reported #8004 :)